### PR TITLE
App/Toponaming: String hasher

### LINKED ISSUE #9127: [Issue] TopoNaming Phase 2 (card 1 of 5) - transfer StringHasher
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Forums discussion

_No response_

### Version

0.20 (Release)

### Full version info

```shell
[code]
OS: Pop!_OS 22.04 LTS (pop:GNOME/pop)
Word size of FreeCAD: 64-bit
Version: 0.20.14555 (Git shallow) Snap 222
Build type: Debug
Branch: master
Hash: d4feb6578398d51eaf56cd794e8d3e2e000bea64
Python 3.10.6, Qt 5.15.3, Coin 4.0.0, Vtk 9.1.0, OCC 7.5.1
Locale: English/United States (en_US)
[/code]
```

### Subproject(s) affected?

None

### Issue description

> The toponaming algorithm needs to hash strings in a specific manner.

## description

Transfer the StringHasher.h/.cpp files to master. Refactor as needed.

For context, see issue #8432

## notes

Used by the ElementMap struct functions.

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by chennes**
This PR will close #9127 -- it is mostly focused on StringHasher, but to pull that code into master it also requires some modifications to Reader, XMLWriter, and Bitmask.

I've marked this as ready for review: there are still a few missing tests related to serialization, but we can consider merging it without those in the interest of getting the other cards moving.

---

- [X]  This Pull Request meets the requirements outlined in section 5 of [CONTRIBUTING.md](https://github.com/FreeCAD/FreeCAD/blob/master/CONTRIBUTING.md) for a Valid PR

### CODE REVIEW

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9148">pipeline status</a> for feature branch [PR_9148](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9148). Pipeline [826752964](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/826752964) was triggered at [94b56d5](https://github.com/FreeCAD/FreeCAD/pull/9148/commits/94b56d522bd9ae0f65268ef376b51a2ea8524001). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9148">pipeline status</a> for feature branch [PR_9148](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9148). Pipeline [826862181](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/826862181) was triggered at [025261e](https://github.com/FreeCAD/FreeCAD/pull/9148/commits/025261eed45a34d6b6270d6e403cdabd98cb239a). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9148">pipeline status</a> for feature branch [PR_9148](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9148). Pipeline [826868415](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/826868415) was triggered at [7bf4c77](https://github.com/FreeCAD/FreeCAD/pull/9148/commits/7bf4c77bdbba39741a62221e92bdbb7821f0d6c9). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9148">pipeline status</a> for feature branch [PR_9148](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9148). Pipeline [829315377](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/829315377) was triggered at [aab82e4](https://github.com/FreeCAD/FreeCAD/pull/9148/commits/aab82e4f1f42262ab8db8d1451bd89120dee9df8). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9148">pipeline status</a> for feature branch [PR_9148](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9148). Pipeline [829421473](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/829421473) was triggered at [6e3be06](https://github.com/FreeCAD/FreeCAD/pull/9148/commits/6e3be060dbe029838274e07e5a527246162933e0). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9148">pipeline status</a> for feature branch [PR_9148](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9148). Pipeline [829522894](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/829522894) was triggered at [1412dce](https://github.com/FreeCAD/FreeCAD/pull/9148/commits/1412dcee341275f92dc8554c07dd742213bd8021). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9148">pipeline status</a> for feature branch [PR_9148](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9148). Pipeline [829528464](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/829528464) was triggered at [a1637a3](https://github.com/FreeCAD/FreeCAD/pull/9148/commits/a1637a30be120be6520b52c36cc4afa2bce8063f). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9148">pipeline status</a> for feature branch [PR_9148](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9148). Pipeline [832286019](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/832286019) was triggered at [37e57d3](https://github.com/FreeCAD/FreeCAD/pull/9148/commits/37e57d3f4fb3a3479b026db6f5a9e51121c415d4). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9148">pipeline status</a> for feature branch [PR_9148](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9148). Pipeline [836956969](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/836956969) was triggered at [d632bc9](https://github.com/FreeCAD/FreeCAD/pull/9148/commits/d632bc90e10608242db51d30a4c771a05ea85005). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9148">pipeline status</a> for feature branch [PR_9148](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9148). Pipeline [838148403](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/838148403) was triggered at [b02c6d5](https://github.com/FreeCAD/FreeCAD/pull/9148/commits/b02c6d59837363e9d6531bd83816949f83ab53fa). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/App/StringHasher.h [hunk 1]
```diff
@@ -0,0 +1,842 @@
...
+    {
+        return *this;
+    };
+};
+
+//////////////////////////////////////////////////////////////////////////
+
+/** Counted reference to a StringID instance
+ */
+class StringIDRef
```
[the remark this answers is not in the document]
Comment (chennes) [reply src/App/StringHasher.h:333]: It maintains `index` as a separate parameter from the rest of the string. It's my understanding that's the real reason for its existence (though @realthunder can confirm, perhaps).

File: src/App/StringHasher.h [hunk 2]
```diff
@@ -0,0 +1,842 @@
...
+private:
+    long _id;
+    QByteArray _data;
+    QByteArray _postfix;
+    StringHasher* _hasher = nullptr;
+    mutable Flags _flags;
+    mutable QVector<StringIDRef> _sids;
+
+private:
+    StringID([[maybe_unused]] const StringID& other)
```
[the remark this answers is not in the document]
Comment (chennes) [reply src/App/StringHasher.h:313]: Yep, I forgot about that option, thanks!

File: src/App/StringHasher.h [hunk 3]
```diff
@@ -0,0 +1,842 @@
...
+     * The input string is expected to be in the format of #<id> or with index
+     * #<id>:<index>, where both id and index are in hex digits.
+     */
+    static IndexID fromString(const QByteArray& bytes, bool eof = true)
+    {
+        return fromString(bytes.constData(), eof, bytes.size());
+    }
+
+    /** Get the text content of this StringID
+     * @param index: optional index
```
[the remark this answers is not in the document]
Comment (chennes) [reply src/App/StringHasher.h:249]: There were a few of these in the original, I missed that one. I've been adding the default parameter.

**Comment by Pesc0** [comment]
I think this can get merged as is. A cleanup/refactor is probably needed on all the work done so far anyway.

File: src/App/StringHasher.h [hunk 4]
```diff
@@ -0,0 +1,842 @@
...
+    void saveStream(std::ostream& stream) const;
+    void restoreStream(std::istream& stream, std::size_t count);
+    void restoreStreamNew(std::istream& stream, std::size_t count);
+
+private:
+    std::unique_ptr<HashMap> _hashes;///< Bidirectional map of StringID and its index (a long int).
+    mutable std::string _filename;
+
+private:
+    StringHasher(const StringHasher&);
```
[the remark this answers is not in the document]
Comment (chennes) [reply src/App/StringHasher.h:788]: Yep.

**Comment by chennes** [comment]
Right - now that we have decent test coverage an ambitious developer can start to refactor the code itself without fear that they will break the functionality. Of course, there is still room for more testing, I didn't catch every possible combination of branches in `getID()`.

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9148">pipeline status</a> for feature branch [PR_9148](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_9148). Pipeline [838436421](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/838436421) was triggered at [da8cce8](https://github.com/FreeCAD/FreeCAD/pull/9148/commits/da8cce8abecd6baa88bd50a781a40450177aa421). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by chennes** [comment]
This is marked as Draft again pending some bug-fix work that @realthunder is working on. It should be ready for merge in a few days.

**Comment by chennes** [comment]
Yes - I'm reviewing it now, and updating the tests.

**Comment by lamikr** [comment]
I have build now the latest version of this pull request succesfully on my Linux distro. (Mageia 8). How can I execute the unit tests? (That seems to fail on Ubuntu at the moment)

**Comment by Pesc0** [comment]
```
./build/tests/Tests_run --gtest_filter=StringIDTest.*:StringIDRefTest.*:StringHasherTest.*:BitmaskTest.*:ReaderTest.*:WriterTest.*
```

**Comment by chennes** [comment]
The tests will fail everywhere at the moment: @realthunder's latest work changes the behavior of the class, so the tests have to get updated to reflect that change. I will get to that Soon.

**Comment by chennes** [comment]
@realthunder I'm working to understand the new behavior after this latest change: given this code
```
    // mappedName set to "Test1;:M;FUS;:Hb:7,F"
    QVector<App::StringIDRef> sids;
    auto inserted = Hasher()->getID(mappedName, sids);

    auto result1 = Hasher()->getID(1);
    auto result2 = Hasher()->getID(2);
    auto result3 = Hasher()->getID(3);

    // Assert
    EXPECT_EQ(";:M;FUS;:Hb:7,F", result1.dataToText());
    EXPECT_EQ("Test1", result2.dataToText());
    EXPECT_EQ("Test1;:M;FUS;:Hb:7,F", result3.dataToText());
```
The second two tests fail, because the 1 is dropped from the string: what's actually in there is "Test" and "Test;:M;FUS;:Hb:7,F". In all three results, _index is zero (though result3 shows itself as both postfixed and indexed in the flags). What's happening to that index value?

**Comment by realthunder** [comment]
The index is stored in `inserted._index`. I think the fact that `getID()` splits the input `MappedName` into three parts should be treated as an internal behavior and may be changed in the future. The external behavior is that `getID()` returns a `StringIDRef` as the reference to the input `MappedName` stored as a text string. The caller can get back the text string with `StringIDRef::dataToText()`.

The reason for stripping the index is so that we do not store names like `Face1` and `Face2` as two unique entries. The current implementation is still not ideal. It would be better If we can find an efficient random accessing (de)compressor to store retrieve the names. And that's also why I say the current splitting behavior should be considered as private.

**Comment by marioalexis84** [comment]
@chennes with this changes I need to `#include <memory>` in `src/Base/Writer.h`  to successful compile on Arch Linux.

**Comment by adrianinsaval** [comment]
Indeed this broke compilation of master on archlinux
```
In file included from /src/Base/Writer.cpp:30:
/src/FreeCAD/src/Base/Writer.h:161:10: error: ‘unique_ptr’ in namespace ‘std’ does not name a template type
  161 |     std::unique_ptr<std::ostream> CharStream;
      |          ^~~~~~~~~~
/src/Base/Writer.h:42:1: note: ‘std::unique_ptr’ is defined in header ‘<memory>’; did you forget to ‘#include <memory>’?
   41 | #include "FileInfo.h"
  +++ |+#include <memory>
   42 | 
```

**Comment by marioalexis84** [comment]
https://github.com/FreeCAD/FreeCAD/pull/9589

**Comment by chennes** [comment]
Thanks, I'll get that merged as soon as the CI finishes.

**Comment by Pesc0** [comment]
@chennes are you aware of these last changes on LinkDaily? 
https://github.com/realthunder/FreeCAD/commit/c35efd8a8b9896ee1dbb2ef6c6720c7f35093ff1
https://github.com/realthunder/FreeCAD/commit/5fb34318b9bb252dc1f70a4a655c1f7a9eada55b
https://github.com/realthunder/FreeCAD/commit/4e4972a09c9c41deb967882d0d5b7d7a5e78332e

**Pull request merged**